### PR TITLE
input: add a pmw3610 driver

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Same drill as the other pixart device, just laying a base here, will followup with some more settings but this seems to work nicely in its current state.

---

Add a driver for the Pixart PMW3610 mouse sensor device.

### CODE REVIEW

File: drivers/input/input_pmw3610.c [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+	ret = gpio_pin_interrupt_configure_dt(&cfg->motion_gpio,
+					      GPIO_INT_EDGE_TO_ACTIVE);
+	if (ret != 0) {
+		LOG_ERR("Motion interrupt configuration failed: %d", ret);
+		return ret;
+	}
+
+	ret = pm_device_runtime_enable(dev);
+	if (ret < 0) {
+		LOG_ERR("Failed to enable runtime power management");
```
Comment (gmarull) [inline drivers/input/input_pmw3610.c:376]: it's alwaus useful to log `ret` value

File: drivers/input/input_pmw3610.c [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+	if (ret < 0) {
+		return;
+	}
+
+	if ((burst_data[BURST_MOTION] & MOTION_STATUS_MOTION) == 0x00) {
+		return;
+	}
+
+	x = ((burst_data[BURST_DELTA_XY_H] << 4) & 0xf00) | burst_data[BURST_DELTA_X_L];
+	y = ((burst_data[BURST_DELTA_XY_H] << 8) & 0xf00) | burst_data[BURST_DELTA_Y_L];
```
Comment (gmarull) [inline drivers/input/input_pmw3610.c:186]: question: same mask for both cases, is this correct?

File: drivers/input/input_pmw3610.c [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+	}
+
+	ret = gpio_pin_interrupt_configure_dt(&cfg->motion_gpio,
+					      GPIO_INT_EDGE_TO_ACTIVE);
+	if (ret != 0) {
+		LOG_ERR("Motion interrupt configuration failed: %d", ret);
+		return ret;
+	}
+
+	ret = pm_device_runtime_enable(dev);
```
Comment (gmarull) [inline drivers/input/input_pmw3610.c:374]: note: when is the device woken up later?

File: drivers/input/input_pmw3610.c (hunk 1 above)
Comment (fabiobaltieri) [reply drivers/input/input_pmw3610.c]: cool, added, gotta fix in in few other drivers too, will followup

File: drivers/input/input_pmw3610.c (hunk 3 above)
Comment (fabiobaltieri) [reply drivers/input/input_pmw3610.c:374]: isn't this going to be controlled by something on the upper layer by calling get/put on them? I may (still) be misunderstanding the runtime pm stuff

File: drivers/input/input_pmw3610.c (hunk 2 above)
Comment (fabiobaltieri) [reply drivers/input/input_pmw3610.c:186]: yeah the mask is after the shift and the shift is different, I'm somewhat surprised that the shift does not push the bits out of the uint8_t but it seems to work, tried

```c
        uint8_t burst_data[4];
        int32_t x, y;
        burst_data[0] = 0x12;
        burst_data[1] = 0x03;
        burst_data[2] = 0x04;
        x = ((burst_data[0] << 4) & 0xf00) | burst_data[1];
        y = ((burst_data[0] << 8) & 0xf00) | burst_data[2];
        printf("%04x %04x\n", x, y);
```

```
0103 0204
```

Comment (aasinclair) [reply drivers/input/input_pmw3610.c:186]: burst_data[0] is implicitly promoted to an int, so this is guaranteed to work.

**Review by aasinclair — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

File: drivers/input/input_pmw3610.c (hunk 2 above)
Comment (fabiobaltieri) [reply drivers/input/input_pmw3610.c:186]: Gotta love this language, always learning something. :-)

Comment (gmarull) [reply drivers/input/input_pmw3610.c:186]: actually, should'nt you use `U` suffixes for the constants? IIRC binary ops are undefined on signed integers

Comment (fabiobaltieri) [reply drivers/input/input_pmw3610.c:186]: is that the case? that's usually something I don't pay too much attention to but I can change it if it does

**Pull request merged**